src/error.rs

```diff
@@ -160,6 +160,10 @@ pub enum Error {
     MetadataDecodeError {
         error: Option<String>,
     },
+
+    GenericFileDiagnostics {
```

Comment:
I think if we pre-render the warnings and put them here we won't be able to have multiple formats for errors later (i.e. LSP, JSON, GitHub Actions diagnostic thingies). The preciseness of the warnings previously worked well for this.

Perhaps an approach like this?

```rust
    let any_warnings = warning::print_all(analysed.as_slice());
    if warnings_as_errors && any_warnings {
        return Err(Error::ForbiddenWarnings);
    }
```
```
error: Warnings detected

The warnings above were found when compiling this project and the `--warnings-as-errors`
flag has been set so we can't continue compilation. Fix the warnings and try again.
```